batch_job/src/batch_job_amazon.c

```diff
@@ -193,6 +309,44 @@ static int aws_terminate_instance( struct aws_config *c, const char *instance_id
 	}
 }
 
+static int idle_instance_expire( struct idle_instance *i, int *timediff )
+{
+	if (difftime(i->last_occupied_time, time(0)) >= *timediff) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+/*
+Recursively terminate expired idle instances based on difftime and timestamp
+*/
+
+static int terminate_expired_instances( struct aws_config *c, const int timediff )
+{
+	const int* timediff_p;
+	timediff_p = &timediff;
+	if (!instance_list){
+		debug(D_BATCH,"idle instance list has not been created\n");
+		return 0;
+	}
+	debug(D_BATCH,"entering aws_terminate_idle_instance, current list count is %d\n", list_size(instance_list));
+	if (list_size(instance_list) == 0){
+		debug(D_BATCH,"idle instance list empty\n");
+		return 0;
+	}
+	struct idle_instance* i = list_find(instance_list, (list_op_t) idle_instance_expire, timediff_p);
+	while (i != NULL){
+		aws_terminate_instance(c, i->instance_id);
+		list_remove(instance_list, i);
+		debug(D_BATCH,"erasing idle instance %s from list \n", i->instance_id);
+		debug(D_BATCH,"exiting terminate_idle_instance, current list count is %d\n", list_size(instance_list));
+		free(i);
```

Comment:
For consistency, create a function called aws_instance_delete() and use it to dispose of an instance, instead of directly using free.  (Also, I think you may need to free members of the object.)